charms/tensorboards-web-app/src/charm.py

```diff
@@ -1,158 +1,252 @@
 #!/usr/bin/env python3
-# Copyright 2021 Canonical Ltd.
+# Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-import logging
+"""A Juju Charm for Tensorboards Web App."""
 
-from oci_image import OCIImageResource, OCIImageResourceError
+import logging
+from typing import Dict
+
+from charmed_kubeflow_chisme.exceptions import ErrorWithStatus, GenericCharmRuntimeError
+from charmed_kubeflow_chisme.kubernetes import KubernetesResourceHandler
+from charmed_kubeflow_chisme.lightkube.batch import delete_many
+from charmed_kubeflow_chisme.pebble import update_layer
+from charms.observability_libs.v1.kubernetes_service_patch import KubernetesServicePatch
+from lightkube import ApiError
+from lightkube.generic_resource import load_in_cluster_generic_resources
+from lightkube.models.core_v1 import ServicePort
 from ops.charm import CharmBase
 from ops.main import main
-from ops.model import ActiveStatus, BlockedStatus, MaintenanceStatus, WaitingStatus
+from ops.model import ActiveStatus, BlockedStatus, Container, MaintenanceStatus, WaitingStatus
+from ops.pebble import Layer
 from serialized_data_interface import NoCompatibleVersions, NoVersionsListed, get_interfaces
 
+K8S_RESOURCE_FILES = ["src/templates/auth_manifests.yaml.j2"]
+LIGHTKUBE_FIELD_MANAGER = "lightkube"
 
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
-
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
-
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
 
-
-class Operator(CharmBase):
+class TensorboardsWebApp(CharmBase):
     def __init__(self, *args):
+        """Initialize charm and setup the container."""
         super().__init__(*args)
 
-        self.log = logging.getLogger(__name__)
-        self.image = OCIImageResource(self, "oci-image")
-
-        for event in [
-            self.on.install,
-            self.on.upgrade_charm,
-            self.on.config_changed,
-            self.on.leader_elected,
-            self.on["ingress"].relation_changed,
-        ]:
-            self.framework.observe(event, self.main)
+        self.logger = logging.getLogger(__name__)
+        self._namespace = self.model.name
+        self._name = self.app.name
+        self._http_port = self.model.config["port"]
```

Comment:
Should the port really be configurable? For starters, looking at the TWA, this looks pretty much [hardcoded](https://github.com/kubeflow/kubeflow/blob/29ee2648f4e0ce597b0fd4ad0b992720aab2e624/components/crud-web-apps/tensorboards/frontend/src/environments/environment.ts#L7) (also [here](https://github.com/kubeflow/kubeflow/blob/29ee2648f4e0ce597b0fd4ad0b992720aab2e624/components/crud-web-apps/tensorboards/Dockerfile#L60) and in the Makefiles), I'm not sure we're doing anything by configuring it here.

Reply:
Note that we removed the port when rewriting the controller as well: https://github.com/canonical/kubeflow-tensorboards-operator/pull/69/commits/6fe2287a4ac94dda04178ea774d64aab880bc65e